single-pool: update docs with examples (#5414)
solana-labs/solana-program-library: docs/src/single-pool.mdx

const rpc = createSolanaRpc({ transport });
```

### Keypair
Web3.js Next uses [`crypto.subtle`](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/subtle) for key management. For local development, one can create an object `{ privateKey, publicKey }` using `crypto.subtle.importKey` to produce signing (private) and verifying (public) keys. Generally, however, one should allow the user wallet to manage keys, as the system when used properly is designed to not allow key material to leave secure storage.

  </TabItem>
</Tabs>

### Setting up a single-validator pool

#### Creating the pool

A single-validator stake pool can be created permissionlessly, by anyone, for a
given vote account. This allows you to receive the full staking yield you would by
staking directly while holding the value in a tokenized form.
It also allows you to buy or sell stakes smaller than the minimum delegation on the market.

Assuming a vote account `Ammgaa2iZfA745BmZMhkcS27uh87fEVDC6Gm2RXz5hrC` exists, we create a pool at address `DkE6XFGbqSyYzRugLVSmmB42F9BQZ7mZU837e2Cti7kb`:

<Tabs className="unique-tabs" groupId="language-selection">
  <TabItem value="cli" label="CLI" default>

```console

> spl-single-pool initialize Ammgaa2iZfA745BmZMhkcS27uh87fEVDC6Gm2RXz5hrC

Reviewer: nit-up-to-you-for-consistency: the earlier examples start the line with `$`, and this one doesn't. maybe it's best to be consistent and not use them anywhere, and instead have an empty line between the command and the output

Author: will add the $ actually, ty for catching this, i copied the above stuff from the token docs

```typescript
const voteAccountAddress = 'Ammgaa2iZfA745BmZMhkcS27uh87fEVDC6Gm2RXz5hrC' as VoteAccountAddress;
const transaction = await SinglePoolProgram.initialize(
  rpc,
  voteAccountAddress,
  feePayerAddress,
);

// sign with the fee payer
```

  </TabItem>
</Tabs>

#### Managing token metadata

By default, when a pool is created, it also creates Metaplex token metadata for the mint associated with the pool.
If, for whatever reason, this was opted out of by the pool creator, anyone may create the default metadata permissionlessly:

<Tabs className="unique-tabs" groupId="language-selection">
  <TabItem value="cli" label="CLI" default>

```console
spl-single-pool create-token-metadata --pool DkE6XFGbqSyYzRugLVSmmB42F9BQZ7mZU837e2Cti7kb
```

  </TabItem>

  <TabItem value="jsc" label="WEB3.JS CLASSIC">

```typescript
const transaction = await SinglePoolProgram.createTokenMetadata(poolAddress, feePayerAddress);

// sign with the fee payer
```

  </TabItem>

  <TabItem value="jsn" label="WEB3.JS NEXT">

```typescript
const transaction = await SinglePoolProgram.createTokenMetadata(poolAddress, feePayerAddress);

// sign with the fee payer
```

  </TabItem>
</Tabs>

> The default token metadata is only minimally helpful, spotlighting the address of the pool.

Reviewer: is this correct?
```suggestion
The default token metadata is only minimally helpful, spotlighting the address of the validator vote account.
```

Author: oops youre right! i was looking at the create with seed call when i wrote this..

The owner of the vote account, however, can change the metadata to anything they wish.
They prove their identity by signing with the vote account's authorized withdrawer;
this is the only permissioned instruction on the pool.

<Tabs className="unique-tabs" groupId="language-selection">
  <TabItem value="cli" label="CLI" default>

```console
spl-single-pool update-token-metadata DkE6XFGbqSyYzRugLVSmmB42F9BQZ7mZU837e2Cti7kb "My Cool Pool" cPool
```

  </TabItem>

  <TabItem value="jsc" label="WEB3.JS CLASSIC">

```typescript
const transaction = await SinglePoolProgram.updateTokenMetadata(
  voteAccountAddress,
  authorizedWithdrawerAddress,
  'My Cool Pool',
  'cPool',
);

// sign with the fee payer and authorized withdrawer
```

  </TabItem>

  <TabItem value="jsn" label="WEB3.JS NEXT">

```typescript
const transaction = await SinglePoolProgram.updateTokenMetadata(
  voteAccountAddress,
  authorizedWithdrawerAddress,
  'My Cool Pool',
  'cPool',
);

// sign with the fee payer and authorized withdrawer
```

  </TabItem>
</Tabs>

> A URL may also be included as an optional parameter.

Reviewer: Could you include the URL in the examples? I think most people will want to do that

```typescript
const transaction = await SinglePoolProgram.deposit({
  rpc,
  pool: poolAddress,
  userWallet,
  userStakeAccount,
});

// sign with the fee payer and stake account withdraw authority, if these signers differ
// userWallet is a convenience parameter to use one account as a payer, authority, and lamport recipient
```

  </TabItem>
</Tabs>

All versions of the `deposit` command/transaction automatically create the associated token account for the pool token if it doesn't exist and no auxiliary token account address is provided.

The program also makes available a convenience address for each pool, called the default deposit address.
This allows a flow where you create and delegate a stake at a program-derived address, and then can deposit this stake after the epoch boundary, without having to generate or keep track of any new keypairs.
The user retains full authority on the stake account until they decide to deposit.

<Tabs className="unique-tabs" groupId="language-selection">
  <TabItem value="cli" label="CLI" default>

```console
spl-single-pool create-default-stake --pool DkE6XFGbqSyYzRugLVSmmB42F9BQZ7mZU837e2Cti7kb 1000000000
```

> (some time passes)

Reviewer: nit: this can be more specific
```suggestion
Once the stake becomes active, typically after crossing one epoch boundary.
```

```console
spl-single-pool deposit --pool DkE6XFGbqSyYzRugLVSmmB42F9BQZ7mZU837e2Cti7kb --default-stake-account
```

  </TabItem>

  <TabItem value="jsc" label="WEB3.JS CLASSIC">

```typescript

> const minimumDelegation = (await connection.getStakeMinimumDelegation()).value;

Reviewer: nit-up-to-you: how about putting a comment that any value greater than or equal to the minimum delegation is valid?

Author: you know, it occurs to me that until the minimum for stake program is activated, they can actually do this with any amount. the only enforcement done anywhere is that the _pool_ doesnt go below max(1sol, minimum). maybe to allay confusion ill just hardcode the number...

const transaction = await SinglePoolProgram.createAndDelegateUserStake(
  connection,
  voteAccountAddress,
  userWallet,
  minimumDelegation,
);

// sign with user wallet
```

> (some time passes)

Reviewer: nit: same here
```suggestion
Once the stake becomes active, typically after crossing one epoch boundary.
```

```typescript
const transaction = await SinglePoolProgram.deposit({
  connection,
  pool: poolAddress,
  userWallet,
  depositFromDefaultAccount: true,
});

// sign with user wallet
```

  </TabItem>

  <TabItem value="jsn" label="WEB3.JS NEXT">

```typescript
const minimumDelegation = (await rpc.getStakeMinimumDelegation().send()).value;
const transaction = await SinglePoolProgram.createAndDelegateUserStake(
  rpc,
  voteAccountAddress,
  userWallet,
  minimumDelegation,
);

// sign with user wallet, which is used as the fee payer and as the base address for a seeded account
```

> (some time passes)

Reviewer: ```suggestion
Once the stake becomes active, typically after crossing one epoch boundary.
```
